rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,226 @@
+  limitations under the License.
+-->
+
+# RFC-56: Federated Storage Layer
+
+## Proposers
+- @umehrot2
+
+## Approvers
+- @vinoth
+- @shivnarayan
+
+## Status
+
+JIRA: [https://issues.apache.org/jira/browse/HUDI-3625](https://issues.apache.org/jira/browse/HUDI-3625)
+
+## Abstract
+
+As you scale your Apache Hudi workloads over Cloud object stores like Amazon S3, there is potential of hitting request
+throttling limits which in-turn impacts performance. In this RFC, we are proposing to support an alternate storage
+layout that is optimized for Amazon S3 and other cloud object stores, which helps achieve maximum throughput and
+significantly reduce throttling.
+
+In addition, we are proposing an interface that would allow users to implement their own custom strategy to allow them
+to distribute the data files across cloud stores, hdfs or on prem based on their specific use-cases.
+
+## Background
+
+Apache Hudi follows the traditional Hive storage layout while writing files on storage:
+- Partitioned Tables: The files are distributed across multiple physical partition folders, under the table's base path.
+- Non Partitioned Tables: The files are stored directly under the table's base path.
+
+While this storage layout scales well for HDFS, it increases the probability of hitting request throttle limits when
+working with cloud object stores like Amazon S3 and others. This is because Amazon S3 and other cloud stores [throttle
+requests based on object prefix](https://aws.amazon.com/premiumsupport/knowledge-center/s3-request-limit-avoid-throttling/).
+Amazon S3 does scale based on request patterns for different prefixes and adds internal partitions (with their own request limits),
+but there can be a 30 - 60 minute wait time before new partitions are created. Thus, all files/objects stored under the
+same table path prefix could result in these request limits being hit for the table prefix, specially as workloads
+scale, and there are several thousands of files being written/updated concurrently. This hurts performance due to
+re-trying of failed requests affecting throughput, and result in occasional failures if the retries are not able to
+succeed either and continue to be throttled.
+
+The traditional storage layout also tightly couples the partitions as folders under the table path. However,
+some users want flexibility to be able to distribute files/partitions under multiple different paths across cloud stores,
+hdfs etc. based on their specific needs. For example, customers have use cases to distribute files for each partition under
+a separate S3 bucket with its individual encryption key. It is not possible to implement such use-cases with Hudi currently.
+
+The high level proposal here is to introduce a new storage layout strategy, where all files are distributed evenly across
+multiple randomly generated prefixes under the Amazon S3 bucket, instead of being stored under a common table path/prefix.
```

Comment:
I think we need to compartmentalize this discussion in 2 actually:

1. Federated Storage support: being able to offload "partitions" to different buckets
2. Addressing the "common prefix" problem

For the latter, solution is not necessary the former, and we already have solution for it -- we eliminate physical partitioning altogether, store tables as non-partitioned one and support "logical-partitioning" on top of it (which requires just 2 ingredients, partition-listing and partitioning constraints on what records are stored in files w/in the partition).

Reply:
Chatted about this w/ @umehrot2 offline: 

 - Logical partitioning is only addressing this issues w/in the scope of a single table, while resorting to StorageStrategy would allow us to also address the throttling issue that might be occurring on the x-table level (when multiple tables are nested under one "datalake" folder)

Reply:
Adding to what @alexeykudinkin mentions, for the "common prefix" problem, the logical partitioning solves the listing performance and partitioning, but during the writing, if all data files are under the same table path without physical partitioning, writing data files (PUT requests) may still get bottlenecked on the throttling as they have the same prefix.  So randomly generated prefixes can still help here.